google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/StreamWriter.java

```diff
@@ -268,26 +296,40 @@ public void flush(long offset) {
    * @throws IOException
    */
   public void refreshAppend() throws IOException, InterruptedException {
-    synchronized (this) {
-      if (shutdown.get()) {
-        LOG.warning("Cannot refresh on a already shutdown writer.");
-        return;
-      }
-      // There could be a moment, stub is not yet initialized.
-      if (clientStream != null) {
-        LOG.info("Closing the stream " + streamName);
-        clientStream.closeSend();
-      }
-      messagesBatch.resetAttachSchema();
-      bidiStreamingCallable = stub.appendRowsCallable();
-      clientStream = bidiStreamingCallable.splitCall(responseObserver);
+    appendAndRefreshAppendLock.lock();
+    if (shutdown.get()) {
+      LOG.warning("Cannot refresh on a already shutdown writer.");
+      appendAndRefreshAppendLock.unlock();
+      return;
+    }
+    // There could be a moment, stub is not yet initialized.
+    if (clientStream != null) {
+      LOG.info("Closing the stream " + streamName);
+      clientStream.closeSend();
     }
+    messagesBatch.resetAttachSchema();
+    bidiStreamingCallable = stub.appendRowsCallable();
+    clientStream = bidiStreamingCallable.splitCall(responseObserver);
     try {
       while (!clientStream.isSendReady()) {
         Thread.sleep(10);
       }
     } catch (InterruptedException expected) {
     }
+    // Currently there is a bug that it took reconnected stream 5 seconds to pick up
+    // stream count. So wait at least 7 seconds before sending a new request.
+    Thread.sleep(
+        Math.max(
+            this.retrySettings.getInitialRetryDelay().toMillis(),
+            Duration.ofSeconds(REFRESH_STREAM_WAIT_TIME).toMillis()));
+    if (this.onSchemaUpdateRunnable != null) {
```

Comment:
With the lock, do we still need this? Since SchemaUpdateRunnable is public interface, I'd like the syntax to be as simple as possible. Having an additional thing about update schema seems to be a little complicated.

Reply:
Removed and created new method refreshConnection